CurriculumAssociates/createjs-accessibility
# Fix tabindices in test app

# Fix tabindices in test app
The test app has been using variables for tabindices that get incremented, which goes against best practices:

> Warning: Avoid using tabindex values greater than 0. Doing so makes it difficult for people who rely on assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)

To reflect best practices, this PR replaces positive tab indices with 0 for focusable elements in the tab order and -1 for elements that are not in the tab order but can be focused programmatically. This PR also fixes tab issues in a few examples, described below.

## Changes

### Remove nextTab in AppWindow.js
All usages of _nextTab in AppWindow.js have been removed and replaced with 0 or -1.

### Fix tab order in Drag and Drop example
In the Drag and Drop example, the tab order would always go from red -> green -> blue draggables no matter their positions on the page because their order in the DOM was not actually updated when they were moved between the starting positions and drop zones.

As an example:
<img width="300" alt="Drag and Drop example with red draggable in green drop zone" src="https://user-images.githubusercontent.com/60579411/140823620-ecff2df3-5cd0-4418-a060-54232af29980.png">

Before: Tab order is red -> green -> blue, which is not good because the red draggable is in a drop zone further down on the page than the green and blue draggables. This happens because the DOM structure stays the same even as the draggable positions change:
1. red draggable (positioned at green drop)
2. green draggable (positioned at start)
3. blue draggable (positioned at start)
4. red drop 
5. green drop
6. blue drop

After: Tab order is green -> blue -> red, which matches the top-to-bottom structure. To achieve this, this PR adds containers at the 3 starting positions, then uses addChild and removeChild on these starting containers and the drop zones to move the draggables to their appropriate position in the DOM:
1. red start
2. green start with child green draggable
3. blue start with child blue draggable
4. red drop 
5. green drop with child red draggable
6. blue drop

Testing:
Checked tab order after each of the following

With mouse:
1. Move from start to nowhere (reverts to start)
2. Move from start to empty drop zone
3. Move from start to occupied drop zone 
4. Move from drop zone to nowhere (reverts to start)
5. Move from drop zone to empty drop zone
6. Move from drop zone to occupied drop zone

With KN:
1. Cancel at start
2. Cancel at drop zone
3. Move from start to empty drop zone
4. Move from start to occupied drop zone 
5. Move from drop zone to same drop zone (reverts to start)
6. Move from drop zone to empty drop zone
7. Move from drop zone to occupied drop zone

### Make clear search button tabbable in Search example
In the Search Example, the x button that clears the search query was not in the tab order because it was not a real button. This PR modifies the Button widget slightly so that it's extendable by ClearButton, which is now in the tab order:
![image](https://user-images.githubusercontent.com/60579411/140956646-eedaf561-caf3-4536-85bd-4ac8ee88ce13.png)

There also seemed to be a bug where the search results would automatically update upon entering text, making the Search button useless, so this PR fixes it so that the search results only update on activating the Search button. Future PRs will fix the other issues around button click behavior and the overlapping text in this example.

Testing:
Clear button shows and hides appropriately
1. Type string
2. Confirm that clear button shows up automatically and is tabbable
3. Delete string with backspace
4. Confirm that clear button disappears automatically

Clear button clears appropriately
1. Type string
2. Press search button with KN to see results
3. Press clear button with KN
4. Confirm that search text and clear button disappear, search results reset
5. Repeat using mouse

### Prevent user from navigating onto elements behind the dialog in Form example
When a dialog is open, elements behind the dialog should not be tabbable. This PR implements this for the Form example, making the form itself not navigable while the dialog opened upon clicking Reset is open.

Testing:
1. Tab through form
2. Press reset button with KN
3. Tab repeatedly to check that nothing behind the dialog is accessible
4. Press cancel button with KN
5. Tab through form to check that all elements are tabbable again
6. Repeat 1-5, but press yes button with KN in step 4 instead

### Added missing spin button and its appropriate tab index
Since a spin button can't have an appropriate tab index if it doesn't exist, this PR adds the missing regY spin button in the Transformation portion of the Color Editor and Transformations example. The tab order of the spin buttons is now correct, but the visual focus indicators are missing, which will be added in a future PR.
